File: Threads Posix/exo10.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h>

#define TAILLE 10
unsigned char pile[TAILLE];
int top = 0; // Indice pour la gestion LIFO (Last-In, First-Out)

sem_t sem_libres, sem_prets;
pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;

void* producteur(void* arg) 
{
    int id = *((int*)arg);
    while (1) 
    {
        unsigned char objet = rand() % 256;
        sleep((rand() % 3) + 1);

        sem_wait(&sem_libres); // Attend une place libre
        pthread_mutex_lock(&mutex);
        
        pile[top++] = objet; // Empile (LIFO)
        printf("[Prod %d] + Déposé: %02X (Stock: %d/%d)\n", id, objet, top, TAILLE);
        
        pthread_mutex_unlock(&mutex);
        sem_post(&sem_prets); // Signale un objet prêt
    }
}

void* consommateur(void* arg) 
{
    int id = *((int*)arg);
    while (1) 
    {
        sleep((rand() % 3) + 1);

        sem_wait(&sem_prets); // Attend un objet
        pthread_mutex_lock(&mutex);
        
        unsigned char objet = pile[--top]; // Dépile (LIFO)
        printf("[Cons %d] - Retiré: %02X (Stock: %d/%d)\n", id, objet, top, TAILLE);
        
        pthread_mutex_unlock(&mutex);
        sem_post(&sem_libres); // Signale une place libérée
    }
}

int main() 
{
    int n, m;
    printf("Nombre de producteurs (N) et consommateurs (M) : ");
    scanf("%d %d", &n, &m);

    sem_init(&sem_libres, 0, TAILLE);
    sem_init(&sem_prets, 0, 0);

    pthread_t prods[n], cons[m];
    int id_p[n], id_c[m];

    for (int i = 0; i < n; i++) 
    { 
        id_p[i] = i; pthread_create(&prods[i], NULL, producteur, &id_p[i]); 
    }
    for (int i = 0; i < m; i++) 
    { 
        id_c[i] = i; pthread_create(&cons[i], NULL, consommateur, &id_c[i]); 
    }

    for (int i = 0; i < n; i++) 
    { 
        pthread_join(prods[i], NULL); 
    }
    for (int i = 0; i < m; i++) 
    {
        pthread_join(cons[i], NULL);
    }

    return 0;
}

File: Threads Posix/exo10.py
import threading
import time
import random

TAILLE = 10
pile = [0] * TAILLE
top = 0

sem_libres = threading.Semaphore(TAILLE)
sem_prets = threading.Semaphore(0)
mutex = threading.Lock()

def producteur(id_prod):
    global top
    while True:
        objet = random.randint(0, 255)
        time.sleep(random.randint(1, 3))

        sem_libres.acquire()
        with mutex:
            pile[top] = objet
            top += 1
            print(f"[Prod {id_prod}] + Déposé: {objet:02X} (Stock: {top}/{TAILLE})")
        sem_prets.release()

def consommateur(id_cons):
    global top
    while True:
        time.sleep(random.randint(1, 3))

        sem_prets.acquire()
        with mutex:
            top -= 1
            objet = pile[top]
            print(f"[Cons {id_cons}] - Retiré: {objet:02X} (Stock: {top}/{TAILLE})")
        sem_libres.release()

def main():
    line = input("Nombre de producteurs (N) et consommateurs (M) : ")
    params = line.split()
    if len(params) < 2: return
    n, m = int(params[0]), int(params[1])

    for i in range(n):
        threading.Thread(target=producteur, args=(i,), daemon=True).start()
    
    for i in range(m):
        threading.Thread(target=consommateur, args=(i,), daemon=True).start()

    try:
        while True: time.sleep(1)
    except KeyboardInterrupt:
        pass

if __name__ == "__main__":
    main()